qiskit/transpiler/passes/scheduling/dynamical_decoupling.py

```diff
@@ -0,0 +1,216 @@
+        circ.cx(1, 2)
+        circ.cx(2, 3)
+        circ.measure_all()
+
+        durations = InstructionDurations(
+            [("h", 0, 50), ("cx", [0, 1], 700), ("reset", None, 10),
+             ("cx", [1, 2], 200), ("cx", [2, 3], 300),
+             ("x", None, 50), ("measure", None, 1000)]
+        )
+
+        dd_sequence = [XGate(), XGate()]
+
+        pm = PassManager([ALAPSchedule(durations),
+                          DynamicalDecoupling(durations, dd_sequence)])
+        circ_dd = pm.run(circ)
+        timeline_drawer(circ_dd)
+    """
+
+    def __init__(self, durations, dd_sequence, qubits=None, spacing=None, skip_reset_qubits=True):
+        """Dynamical decoupling initializer.
+
+        Args:
+            durations (InstructionDurations): Durations of instructions to be
+                used in scheduling.
+            dd_sequence (list[Gate]): sequence of gates to apply in idle spots.
+            qubits (list[int]): physical qubits on which to apply DD.
+                If None, all qubits will undergo DD (when possible).
+            spacing (callable): a function that specifies spacing between DD
+                gates. It maps natural numbers, i, to the fraction of the total
+                slack that must be allocated to the i'th delay window. If None,
+                equal spacing will be used.
+            skip_reset_qubits (bool): if True, does not insert DD on idle
+                periods that immediately follow initialized/reset qubits (as
+                qubits in the ground state are less susceptile to decoherence).
+        """
+        super().__init__()
+        self._durations = durations
+        self._dd_sequence = dd_sequence
+        self._qubits = qubits
+        self._spacing = spacing
+        self._skip_reset_qubits = skip_reset_qubits
+
+    def run(self, dag):
+        """Run the DynamicalDecoupling pass on dag.
+
+        Args:
+            dag (DAGCircuit): a scheduled DAG.
+
+        Returns:
+            DAGCircuit: equivalent circuit with delays interrupted by DD,
+                where possible.
+
+        Raises:
+            TranspilerError: if the circuit is not mapped on physical qubits.
+        """
+        if len(dag.qregs) != 1 or dag.qregs.get("q", None) is None:
+            raise TranspilerError("DD runs on physical circuits only.")
+
+        if dag.duration is None:
+            raise TranspilerError("DD runs after circuit is scheduled.")
+
+        # we support only even number of DD pulses (or just 1, as a special case)
+        if len(self._dd_sequence) != 1:
+            if len(self._dd_sequence) % 2 != 0:
+                raise TranspilerError("DD sequence must contain an even number of gates (or 1).")
+            noop = Operator(IGate())
+            for gate in self._dd_sequence:
+                noop = noop.compose(Operator(gate))
+            if not noop.equiv(IGate()):
+                raise TranspilerError("The DD sequence does not make an identity operation.")
```

Comment:
For better performance this could be replaced by:

```suggestion
            noop = np.eye(2)
            for gate in self._dd_sequence:
                noop = noop.dot(gate.to_matrix())
            if not np.allclose(noop, np.eye(2), rtol=1e-5, atol=1e-8):
                raise TranspilerError("The DD sequence does not make an identity operation.")
```